mouse.lisp

```diff
@@ -0,0 +1,124 @@
+           (min-y (frame-display-y group frame))
+           (max-x (+ min-x (frame-width frame)))
+           (max-y (+ min-y (frame-display-height group frame))))
+      (and (<= min-x mouse-x max-x)
+           (<= min-y mouse-y max-y)))))
+
+(defun mouse-banish-frame (frame)
+  "Banish mouse to corner of frame"
+  (let* ((group (current-group))
+         (min-x (frame-x frame))
+         (max-x (+ min-x (frame-width frame)))
+         (new-x (if (minusp *mouse-follow-banish-x-offset*)
+                    (+ max-x *mouse-follow-banish-x-offset*)
+                    (+ min-x *mouse-follow-banish-x-offset*)))
+         (min-y (frame-display-y group frame))
+         (max-y (+ min-y (frame-display-height group frame)))
+         (new-y (if (minusp *mouse-follow-banish-y-offset*)
+                    (+ max-y *mouse-follow-banish-y-offset*)
+                    (+ min-y *mouse-follow-banish-y-offset*))))
+    (ratwarp (clamp new-x min-x max-x)
+             (clamp new-y min-y max-y))))
+
+(defgeneric mouse-inside-window-p (window)
+  (:documentation "Determine if mouse already inside window.")
+  (:method ((window float-window))
+    (multiple-value-bind (mouse-x mouse-y)
+        (xlib:global-pointer-position *display*)
+      (let* ((leniency-offset 2)
+             (x (xlib:drawable-x (window-parent window)))
+             (w (xlib:drawable-width (window-parent window)))
+             (min-x (- x *float-window-border* leniency-offset))
+             (max-x (+ x w *float-window-border* leniency-offset))
+             (y (xlib:drawable-y (window-parent window)))
+             (h (xlib:drawable-height (window-parent window)))
+             (min-y (- y *float-window-title-height* leniency-offset))
+             (max-y (+ y h *float-window-border* leniency-offset)))
+        (and (<= min-x mouse-x max-x)
+             (<= min-y mouse-y max-y)))))
+  (:method ((window tile-window))
+    (let ((frame (window-frame window)))
+      (mouse-inside-frame-p frame))))
+
+(defun mouse-banish-window (window)
+  "Move mouse pointer to edge of a window."
+  (let* ((min-x (xlib:drawable-x (window-parent window)))
+         (max-x (+ min-x (xlib:drawable-width (window-parent window))))
+         (new-x (if (minusp *mouse-follow-banish-x-offset*)
+                    (+ max-x *mouse-follow-banish-x-offset*)
+                    (+ min-x *mouse-follow-banish-x-offset*)))
+         (min-y (xlib:drawable-y (window-parent window)))
+         (max-y (+ min-y (xlib:drawable-height (window-parent window))))
+         (new-y (if (minusp *mouse-follow-banish-y-offset*)
+                    (+ max-y *mouse-follow-banish-y-offset*)
+                    (+ min-y *mouse-follow-banish-y-offset*))))
+    (ratwarp (clamp new-x min-x max-x)
+             (clamp new-y min-y max-y))))
+
+(defun mouse-handle-focus-frame (current-frame last-frame)
+  "Move mouse when moving frames."
+  (when (eq *mouse-follow-policy* :follow)
+    (unless (or (eq current-frame last-frame)
```

Comment:
We should avoid nesting if when and unless and move the conditions to the outer when.

```
(when (and (eq *mouse-follow-policy* :follow)
           ;; Although frames should be compared by EQUALP afaiu
           (not (eq current-frame last-frame))
           (mouse-outside-frame-p current-frame))) ; or (not (mouse-inside ..))
```

Ideally move the predicate to a function that conveys their intent. Maybe something like

```lisp
(defun mouse-outside-of-old-frame (current-frame last-frame)
  (and (not (eq current-frame last-frame))
       (not (mouse-inside-frame-p current-frame))))
```

(The name is just the first thing I came up with, if you can please come up with a better name for the predicate that communicate what we want to find out)